File: commands/reserves/lend_reserves.py
from argparse import Namespace, _SubParsersAction

from world import load_world, save_world


def run(args: Namespace) -> None:
    world = load_world()

    central_bank_id = args.central_bank_id
    bank_id = args.bank_id
    amount = args.amount

    if amount <= 0:
        print("Amount must be greater than 0.")
        return

    if central_bank_id not in world["central_banks"]:
        print(f"Central bank does not exist: {central_bank_id}")
        return

    if bank_id not in world["banks"]:
        print(f"Bank does not exist: {bank_id}")
        return

    central_bank = world["central_banks"][central_bank_id]
    bank = world["banks"][bank_id]

    central_bank["reserve_accounts"][bank_id] = (
        central_bank["reserve_accounts"].get(bank_id, 0) + amount
    )

    central_bank["loans_to_banks"][bank_id] = (
        central_bank["loans_to_banks"].get(bank_id, 0) + amount
    )

    bank["reserve_balances"][central_bank_id] = (
        bank["reserve_balances"].get(central_bank_id, 0) + amount
    )

    bank["loans_from_central_banks"][central_bank_id] = (
        bank["loans_from_central_banks"].get(central_bank_id, 0) + amount
    )

    save_world(world)

    currency = central_bank["currency"]

    print(f"Lent {amount} {currency} reserves from {central_bank_id} to {bank_id}.")
    print(
        f"{central_bank_id} reserve account for {bank_id}: {central_bank['reserve_accounts'][bank_id]} {currency}"
    )
    print(
        f"{central_bank_id} loan to {bank_id}: {central_bank['loans_to_banks'][bank_id]} {currency}"
    )
    print(
        f"{bank_id} reserves at {central_bank_id}: {bank['reserve_balances'][central_bank_id]} {currency}"
    )
    print(
        f"{bank_id} loan from {central_bank_id}: {bank['loans_from_central_banks'][central_bank_id]} {currency}"
    )


def register(subparsers: _SubParsersAction) -> None:
    parser = subparsers.add_parser(
        "lend-reserves",
        help="Create central bank reserves by lending them to a commercial bank.",
    )

    parser.add_argument("central_bank_id", help="Central bank ID, for example: ecb")

    parser.add_argument("bank_id", help="Commercial bank ID, for example: bank1")

    parser.add_argument("amount", type=int, help="Amount of reserves to lend.")

    parser.set_defaults(func=run)

File: go/internal/commands/reserves/lend_reserves.go
package reserves

import (
	"fmt"
	"strconv"
	"strings"

	"mini-world-go/internal/commandlog"
	"mini-world-go/internal/commandrun"
	"mini-world-go/internal/domain"
	"mini-world-go/internal/world"

	"github.com/spf13/cobra"
)

func newLendReservesCommand() *cobra.Command {
	var collateralAssetID string

	cmd := &cobra.Command{
		Use:   "lend-reserves central_bank_id bank_id currency amount --collateral asset_id",
		Short: "Central bank lends reserves against collateral.",
		Args:  cobra.ExactArgs(4),
		RunE: func(cmd *cobra.Command, args []string) error {
			return runLendReserves(args, collateralAssetID)
		},
	}

	cmd.Flags().StringVar(&collateralAssetID, "collateral", "", "Collateral asset ID.")
	_ = cmd.MarkFlagRequired("collateral")

	return cmd
}

func runLendReserves(args []string, collateralAssetID string) error {
	centralBankID := args[0]
	bankID := args[1]
	currency := strings.ToUpper(args[2])
	amount, err := strconv.Atoi(args[3])
	if err != nil {
		return fmt.Errorf("amount must be an integer")
	}

	w, err := world.Load()
	if err != nil {
		return err
	}

	reserveLoanID, err := domain.LendReserves(w, centralBankID, bankID, currency, amount, collateralAssetID)
	if err != nil {
		return commandrun.PrintBusinessError(err)
	}

	if err := commandrun.SaveWithHistory(w, "lend-reserves", append(args, "--collateral", collateralAssetID)); err != nil {
		return err
	}

	centralBank := w.CentralBanks[centralBankID]
	bank := w.Banks[bankID]
	reserveLoan := w.ReserveLoans[reserveLoanID]
	commandlog.Action("Lent %d %s reserves from %s to %s", amount, currency, centralBankID, bankID)
	commandlog.State("Reserve loan: %s", reserveLoanID)
	commandlog.State("Collateral: %s", collateralAssetID)
	commandlog.State("%s reserve account for %s: %d %s", centralBankID, bankID, centralBank.ReserveAccounts[bankID], currency)
	commandlog.State("%s reserves at %s: %d %s", bankID, centralBankID, bank.ReserveBalances[centralBankID], currency)
	commandlog.State("%s loan to %s: %d %s", centralBankID, bankID, centralBank.LoansToBanks[bankID], currency)
	commandlog.State("%s loan from %s: %d %s", bankID, centralBankID, bank.LoansFromCentralBanks[centralBankID], currency)
	commandlog.State("%s outstanding: %d %s", reserveLoanID, reserveLoan.Outstanding, currency)
	return nil
}
